exp/api/v1beta1/ipam_types.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+const (
+	// IPPoolLabel is used to add the pool name to an IPAddressClaim and IPAddress.
+	IPPoolLabel = "ipam.cluster.x-k8s.io/pool"
+)
+
+// IPAddressSpec is the desired state of an IPAddress.
+type IPAddressSpec struct {
+	// ClaimRef is a reference to the claim this IPAddress was created for.
+	ClaimRef corev1.LocalObjectReference `json:"claim,omitempty"`
+
+	// PoolRef is a reference to the pool that this IPAddress was created from.
+	PoolRef corev1.TypedLocalObjectReference `json:"pool,omitempty"`
```

Comment:
TypedLocalObjectReference feels a bit awkward

```go
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
```

This means apiGroup is optional, but in our case it's never optional? (as we won't have IPPools in the core CAPI group)

Reply:
I personally would prefer local types over types which don't fit

Reply:
And I thought I managed to navigate my way around this discussion :smile:

Reply:
Sorry :)

Reply:
> we could even have a breaking change (this is experimental api after all) to move back to apiVersion.

Not sure what exactly our rules are for beta experimental API types. I would have assumed that the following applies:
> MUST remain backwards compatible unless updates are coinciding with a breaking Cluster API release
> from CONTRIBUTING.md > Experiments > Beta-stage experiments (assume v1beta1 =>  beta stage experiment)

But I think in general I think we have some leeway here in v1beta1. At some point there will be hopefully a decision on #6539 and we can then change the type later on the same way we do it for our core types (with/without apiVersion bump). The situation will be just the same as with our clusterv1 API types.

Reply:
Given that we are following the same pattern as for our clusterv1 types, I'll resolve this conversation